vw/src/main/scala/com/microsoft/azure/synapse/ml/vw/VowpalWabbitBaseProgressive.scala

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.azure.synapse.ml.vw
+
+import org.apache.spark.TaskContext
+import org.apache.spark.ml.Transformer
+import org.apache.spark.sql.{DataFrame, Dataset, Row}
+import org.apache.spark.sql.catalyst.encoders.RowEncoder
+import org.apache.spark.sql.types.{StructType}
+import org.vowpalwabbit.spark.VowpalWabbitNative
+
+import java.util.UUID
+
+trait VowpalWabbitBaseProgressive extends Transformer with VowpalWabbitBase {
+  class TrainingPartitionIterator(inputRows: Iterator[Row],
+                                  localInitialModel: Option[Array[Byte]],
+                                  synchronizationSchedule: VowpalWabbitSynchronizationSchedule,
+                                  vwArgs: String,
+                                  numTasks: Int)
+    extends Iterator[Row] {
+
+      var closed = false
+
+    // TODO: make sure we don't do double close
+    override def finalize(): Unit = close
+
+    private def close(): Unit = {
+      if (!closed) {
+        vw.close()
+        closed = true
+      }
+    }
+
+    // note this is executed on each worker
+    lazy val vw = {
+      val contextArgs = if (numTasks == 1) "" else s"--node ${TaskContext.get.partitionId}"
+
+      val args = buildCommandLineArguments(vwArgs, contextArgs)
+
+      if (localInitialModel.isEmpty) new VowpalWabbitNative(args)
+      else new VowpalWabbitNative(args, localInitialModel.get)
+    }
+
+    override def hasNext: Boolean = {
+      val ret = inputRows.hasNext
+
+      if (!ret) {
+        // blocks until sync in AllReduce use-case is done
+        // make sure that all partitions call endPass the same amount of time
+        // (and at least once at the end)
+        0 to synchronizationSchedule.getFinishTriggerCount() foreach { _ =>
+//          println(s"AllReduce Finalize ${TaskContext.getPartitionId()}")
+          vw.endPass()
+        }
+
+        // cleanup
+        close()
+      }
+
+      ret
+    }
+
+    override def next(): Row = {
+      try {
+        val inputRow = inputRows.next()
+
+        0 until synchronizationSchedule.getAllReduceTriggerCount(inputRow) foreach { _ =>
+          vw.endPass()
+        }
+
+        // withColumn
+        Row.fromSeq(inputRow.toSeq :+ trainFromRow(vw, inputRow))
+      }
+      catch {
+        case e: Exception =>
```

Comment:
seems like a broad catch with a final exceptiont hat is not typed specifically enough. Perhaps throw something like a RuntimeException

Reply:
I actually don't care which exception comes, I'll just rethrow. I need to invoke close in case of error. should I use error instead?